etcd-backup/main.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/spf13/cobra"
+	"go.etcd.io/etcd/client/pkg/v3/transport"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/client/v3/snapshot"
+
+	"go.uber.org/zap"
+)
+
+const (
+	DefaultDialTimeout       = 15 * time.Second
+	DefaultEtcdClientTimeout = 5 * time.Minute
+)
+
+type options struct {
+	clusterID      string
+	etcdEndpoint   string
+	s3BucketName   string
+	s3BucketRegion string
+}
+
+func main() {
+	var opts options
+
+	cmd := &cobra.Command{
+		Use:          "etcd-backup",
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return run(opts)
+		},
+	}
+
+	cmd.Flags().StringVar(&opts.clusterID, "cluster-id", "", "ID of the hosted cluster, used as a key for the snapshot file in S3")
+	cmd.Flags().StringVar(&opts.etcdEndpoint, "etcd-endpoint", "", "endpoint of the etcd cluster to backup.")
+	cmd.Flags().StringVar(&opts.s3BucketName, "s3-bucket-name", "", "name of the S3 bucket to store etcd backups.")
+	cmd.Flags().StringVar(&opts.s3BucketRegion, "s3-bucket-region", "", "AWS region of the S3 bucket to store etcd backups.")
+
+	cmd.MarkFlagRequired("cluster-id")
+	cmd.MarkFlagRequired("etcd-endpoint")
+	cmd.MarkFlagRequired("s3-bucket-name")
+
+	if err := cmd.Execute(); err != nil {
+		os.Exit(1)
```

Comment:
Wouldn't it be better to have different numbers per different types of errors?

Reply:
hmm, I think the convention is to just indicate an error with `1` and print an error message.
otherwise we would need to document what the different error codes mean.